File: fs/Stone.fs
namespace Domino
// type StoneObj = {
//   id: string
//   turned: Turned
//   left: int
//   right: int
// }

type Stone = class
  val id : string
  val turned : Domino.Turned
  val left : int
  val right : int

  new (id0: string, turned0: Turned, left0: int, right0: int) = {
    id = id0;
    turned = turned0;
    left = left0;
    right = right0
  }

  member this.hasNumber (n: int) = n.Equals(this.left) || n.Equals(this.right);

  member this.isStone (left: int, right: int) =
    (this.left.Equals(left) && this.right.Equals(right)) ||
    (this.left.Equals(right) && this.right.Equals(left))

  member this.turnTo(turns: (Stone) -> Stone) = turns(this)

  member this.toString = sprintf "%d|%d" this.left this.right

  member this.idString =
    if this.left < this.right then
      sprintf "%d|%d" this.left this.right
    else
      sprintf "%d|%d" this.right this.left

  member this.equal(o: Stone) = this.idString.Equals(o.idString)

  // member this.asObj() y StoneObj = {
  //     id: this.id,
  //     turned: this.turned,
  //     left: this.left,
  //     right: this.right,
  // }

end

File: src/stone.ts
import { Turned } from './turned';

export interface StoneObj {
  readonly id: string;
  readonly turned: Turned;
  readonly left: number;
  readonly right: number;
}

export class Stone implements StoneObj {
  public readonly id: string;
  public readonly turned: Turned;
  public readonly left: number;
  public readonly right: number;

  public static create(id: string, turned: Turned, left: number, right: number) {
    return new Stone(id, turned, left, right);
  }

  public static fromTuple([id, left, right]: [string, number, number]): Stone {
    return new Stone(id, Turned.NO, left, right)
  }

  private constructor(id: string, turned: Turned, left: number, right: number) {
    this.left = left;
    this.right = right;
    this.turned = turned;
    this.id = id;
  }
  public hasNumber(n: number): boolean {
    return n == this.left || n == this.right;
  }

  public isStone(left: number, right: number) {
    return (this.left == left && this.right == right) || (this.left == right && this.right == left);
  }

  public turnTo(turns: (my: Stone) => Stone) {
    return turns(this);
  }

  public toString() {
    return `${this.left}|${this.right}`;
  }

  public toTuple(): [string, number, number] {
    return [this.id, this.left, this.right];
  }

  public idString() {
    if (this.left < this.right) {
      return `${this.left}|${this.right}`;
    } else {
      return `${this.right}|${this.left}`;
    }
  }

  public equal(o: Stone) {
    return this.idString() === o.idString();
  }

  public get asObj(): StoneObj {
    return {
      id: this.id,
      turned: this.turned,
      left: this.left,
      right: this.right,
    };
  }
}
